chasm/tree.go

```diff
@@ -137,6 +152,378 @@ func NewEmptyTree(
+	nodeValueT := reflect.TypeOf(n.value)
+	nodeValueV := reflect.ValueOf(n.value)
+
+	protoMessageFound := false
+	for i := 0; i < nodeValueT.Elem().NumField(); i++ {
+		fieldV := nodeValueV.Elem().Field(i)
+		fieldT := fieldV.Type()
+		if !fieldT.AssignableTo(protoMessageT) {
+			continue
+		}
+
+		if fieldV.IsNil() {
+			continue
+		}
+
+		if protoMessageFound {
+			return serviceerror.NewInternal("only one proto field allowed in component")
+		}
+		protoMessageFound = true
+
+		blob, err := serialization.ProtoEncodeBlob(fieldV.Interface().(proto.Message), enumspb.ENCODING_TYPE_PROTO3)
+		if err != nil {
+			return err
+		}
+
+		n.serializedValue.GetComponentAttributes().Data = blob
+		n.serializedValue.GetComponentAttributes().Type = nodeValueT.Elem().String() // type name w/o *
+		if n.serializedValue.GetInitialVersionedTransition() == nil {
+			n.serializedValue.InitialVersionedTransition = &persistencespb.VersionedTransition{
+				TransitionCount:          n.backend.NextTransitionCount(),
+				NamespaceFailoverVersion: n.backend.GetCurrentVersion(),
+			}
+		}
+	}
+	return nil
+}
+
+// Sync tree starting from node n with underlining component value:
+//   - if a child is nil in a corresponding component, it is removed from the tree,
+//   - if a child is no longer in a corresponding component, it is removed from the tree,
+//   - when a child is removed, all its children are removed too.
+//
+// Returns slice of paths to removed nodes.
+func (n *Node) syncChildren() ([]string, error) {
+	var removedPaths []string
+	err := n.syncChildrenInternal(&removedPaths, nil)
+	return removedPaths, err
+}
+
+func (n *Node) syncChildrenInternal(
+	removedPaths *[]string,
+	nodePath []string,
+) error {
+	nodeValueT := reflect.TypeOf(n.value)
+	nodeValueV := reflect.ValueOf(n.value)
+
+	childrenToKeep := make(map[string]struct{})
+	for i := 0; i < nodeValueT.Elem().NumField(); i++ {
+		fieldV := nodeValueV.Elem().Field(i)
+		fieldT := fieldV.Type()
+
+		if fieldT.Kind() != reflect.Ptr {
+			continue
+		}
+
+		if fieldV.IsNil() {
+			continue
+		}
+
+		fieldN := fieldName(nodeValueT.Elem().Field(i))
+
+		switch genericTypePrefix(fieldT.Elem()) {
+		case chasmFieldTypePrefix:
+			//nolint:revive // Internal field is guaranteed to be of type fieldInternal.
+			internal := fieldV.Elem().FieldByName(internalFieldName).Interface().(fieldInternal)
+			if internal.node == nil {
+				// This can't happen!
+				// If parent is in the tree all children should be in the tree too i.e. has node set.
+				// TODO: use softassert
+				panic("internal.node is nil")
+			}
+			if err := internal.node.syncChildrenInternal(removedPaths, append(nodePath, fieldN)); err != nil {
+				return err
+			}
+
+			childrenToKeep[fieldN] = struct{}{}
+		case chasmCollectionTypePrefix:
+			childrenToKeep[fieldN] = struct{}{}
+			// TODO: need to go over every item in collection and update children for it.
+			panic("not implemented")
+		}
+	}
+
```

Comment:
What about newly added fields? when will we add them to the child map?

Is that logic `reflectSubcomponents` or setValues()? I guess I am a bit confused where/when those two methods will be invoked.